Migrate JsonConverter to V2

EricBorczuk opened the pull request:
<!--  Thanks for sending a pull request!  Before submitting:

1. Read our CONTRIBUTING.md guide
2. Rebase your PR if it gets out of sync with main
-->

**What this PR does**:
Migrates JsonConverter (with testing) to sgv2-docsapi

**Which issue(s) this PR fixes**:
Fixes #1727 

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [ ] Documentation added/updated
- [x] CLA Signed: [DataStax CLA](https://cla.datastax.com/)

EricBorczuk commented on sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/service/json/JsonConverter.java:
```diff
@@ -0,0 +1,375 @@
+  }
+
+  public JsonNode convertToJsonDoc(
+      List<QueryOuterClass.Row> rows,
```
Now, instead of providing a list of datastore rows, you must provide a list of grpc rows, along with a list of columns that can be used to determine positional access in the rows.

olim7t replied:
In #1796 I introduced an `ExtendedRow` (name subject to change) wrapper that holds the column definitions, maybe it could be useful here too.

tatu-at-datastax commented on sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/service/json/JsonConverter.java:
```diff
@@ -0,0 +1,375 @@
+
+  @Inject
+  public JsonConverter(ObjectMapper mapper, DocumentProperties docsProperties) {
+    if (mapper == null) {
```
Works, although could also use:

    Objects.requireNonNull(mapper, "JsonConverter requires a non-null ObjectMapper");

EricBorczuk commented:
This should be ready for review now - the DeadLeafCollectors just got added in. As far as firing off the Dead Leaf collection events to grpc, I'm working on that in a separate PR

ivansenic commented on sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/service/json/JsonConverter.java:
```diff
@@ -0,0 +1,364 @@
+
+        if (isArray) {
+          // This removes leading zeros if applicable
+          p = "[" + Integer.parseInt(p.substring(1, p.length() - 1)) + "]";
```
this could be part of the docs api utils?

EricBorczuk replied:
I haven't ported that class yet...is there already a place where it's been partially ported or not yet?

ivansenic commented on sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/service/json/JsonConverter.java:
```diff
@@ -0,0 +1,364 @@
+        ((ObjectNode) ref).set(leaf, n);
+      } else if (ref.isArray()) {
+        if (!leaf.startsWith("["))
+          throw new RuntimeException("Trying to write object to array " + leaf);
```
should we change this and other exceptions with the `ErrorCodeRuntimeException`?

EricBorczuk replied:
This really should be an `IllegalStateException`, as it should never occur normally

ivansenic commented on sgv2-docsapi/src/test/java/io/stargate/sgv2/docsapi/api/service/json/JsonConverterTest.java:
```diff
@@ -0,0 +1,696 @@
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class JsonConverterTest {
```
could this be a `@QuarkusTest` where you simply inject the `JsonConverter` and actually test with respect to the setup data store and document properties, created object mapper, etc.. You can also adapt properties if needed, for example `stargate.document.max-dept=6`. You can see other tests to figure out how the setup a dedicated config for the test?

ivansenic reviewed (changes requested):
It's very hard to review this as this was not refactored in the V1.. It's a lot of cases and following the code is very hard.. Not sure if we should invest more time in improving it, or simply create a ticket for this.. But me understanding what's going on there is impossible..

Left some meaningful comments, but still the major issue is that very complex `convertToJsonDoc` method..

EricBorczuk commented on sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/service/json/DeadLeaf.java:
```diff
@@ -0,0 +1,13 @@
+
+@Value.Immutable(prehash = true)
+public interface DeadLeaf {
+  String STAR = "*";
+  String ARRAY = "[*]";
```
This is related directly to dead leaf stuff - I can put it in Constants though, yeah

ivansenic replied:
As you wish..

EricBorczuk commented on sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/service/json/DeadLeafCollectorImpl.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.Map;
+import java.util.Set;
+
+public class DeadLeafCollectorImpl implements DeadLeafCollector {
```
I think the Immutable was only used to create empty no-op DeadLeafCollectors - which would mean the Immutable is actually not necessary

EricBorczuk commented:
I addressed most of the review, but I guess this PR should wait until after https://github.com/stargate/stargate/pull/1796 is merged (so that I can use the extended row wrapper)?

ivansenic commented on sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/service/json/JsonConverter.java:
```diff
@@ -0,0 +1,364 @@
+   * @param columnSpecs The gRPC column specs
+   * @return A Map of column name to the index of the column.
+   */
+  public Map<String, Integer> generateColumnIndexMap(List<QueryOuterClass.ColumnSpec> columnSpecs) {
```
It's now in the wrapper..

ivansenic commented:
> I addressed most of the review, but I guess this PR should wait until after #1796 is merged (so that I can use the extended row wrapper)?

Yup let's rebase this..

Regarding the new wrapper, it's actually wrapping a single row with columns specs.. But I guess we need more rows.. Not sure if we did this correctly, maybe the `RowWrapper` should be the `RowsWrapper`.. Anyway we can optimize that later, fine to go with `List<RowWrapper>` for now..

ivansenic commented on sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/service/json/JsonConverter.java:
```diff
@@ -0,0 +1,364 @@
+
+    for (QueryOuterClass.Row row : rows) {
+      Long rowWriteTime = row.getValues(columnNameToIndex.get("writetime(leaf)")).getInt();
+      String rowLeaf = row.getValues(columnNameToIndex.get("leaf")).getString();
```
Yea I would also do this after pending, the `documentProperties#tableProperties()` have all the names, plus the function to get path columns.. Let's add the one for the `writetime` as well (default impl based on the leaf column name)..

and the wrapper would simplify this to:

`rowWrapper.getString(tableProperties().stringValueColumnName());`

ivansenic commented on sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/service/common/model/RowWrapper.java:
```diff
@@ -66,6 +66,22 @@ default Double getDouble(String columnName) {
+   * @return Value of the column as int.
+   * @throws IllegalArgumentException If column does not exist.
+   */
+  default Integer getInt(String columnName) {
+    return Values.int_(getValue(columnName));
+  }
```
Don't think we need this, the number booleans use `TINYINT`.

ivansenic commented on sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/service/common/model/RowWrapper.java:
```diff
@@ -66,6 +66,22 @@ default Double getDouble(String columnName) {
+   * @return Value of the column as long.
+   * @throws IllegalArgumentException If column does not exist.
+   */
+  default Long getLong(String columnName) {
+    return Values.bigint(getValue(columnName));
+  }
```
Can you extend the RowWrapperTest to add a quick test for this method as well?

ivansenic commented on sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/service/json/JsonConverter.java:
```diff
@@ -0,0 +1,378 @@
+    }
+
+    for (RowWrapper row : rows) {
+      Long rowWriteTime = row.getLong("writetime(leaf)");
```
Mentioned before `"writetime(leaf)"` should be existing in the `TableProperties` as the default method.

ivansenic commented on sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/service/json/JsonConverter.java:
```diff
@@ -0,0 +1,378 @@
+
+    for (RowWrapper row : rows) {
+      Long rowWriteTime = row.getLong("writetime(leaf)");
+      String rowLeaf = row.getString("leaf");
```
Use `documentProperties.tableProperties().leafColumnName()`

ivansenic commented on sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/service/json/JsonConverter.java:
```diff
@@ -0,0 +1,378 @@
+   * is an Array.
+   */
+  private JsonNode changeCurrentNodeToArray(RowWrapper row, JsonNode parentRef, int pathIndex) {
+    String pbefore = row.getString("p" + (pathIndex - 1));
```
Use `documentProperties.tableProperties().pathColumnName(index)`

ivansenic reviewed (changes requested):
Some things still missing.. The rebase did not update the name of the columns to be read from the documentProperties.

ivansenic commented on sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/common/properties/document/DocumentTableProperties.java:
```diff
@@ -49,4 +49,9 @@ default String pathColumnName(int index) {
+
+  /** @return The name of the column for the writetime of a row */
+  default String writetimeColumnName() {
+    return "writetime(leaf)";
```
```
    return "writetime(" + leafColumnName() + ")":
```

:nerd_face:

ivansenic commented on sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/service/json/JsonConverter.java:
```diff
@@ -0,0 +1,383 @@
+        n = new TextNode(value);
+      }
+    } else if (!row.isNull(docsProperties.tableProperties().booleanValueColumnName())) {
+      Boolean booleanFromRow;
+      if (numericBooleans) {
+        booleanFromRow =
+            (row.getByte(docsProperties.tableProperties().booleanValueColumnName()) != 0);
+      } else {
+        booleanFromRow = row.getBoolean(docsProperties.tableProperties().booleanValueColumnName());
+      }
+
+      n = BooleanNode.valueOf(booleanFromRow);
```
So what's the reason not to go for:

```
DocsApiUtils#getBooleanFromRow(row, documentProperties, numericBooleans)
```

This way we reuse the same logic everywhere.. In fact, it would be way easier to read values before if and switch:

```
String stringValue = DocsApiUtils.getStringFromRow(row, documentProperties);
Boolean booleanValue = DocsApiUtils.getBooleanFromRow(row, documentProperties, numericBooleans);
String doubleValue = DocsApiUtils.getDoubleFromRow(row, documentProperties);

if (stringValue != null) {
   ...
} else if (booleanValue != null) {
  ...

} else if (doubleValue != null) {
  ...
}
```

EricBorczuk replied:
Ah yea, I should've done this - my bad. idk how I didn't notice to do this

Will fix this

ivansenic reviewed (approved):
Can you please the two remaining comments before merging.. I hope you'll find them as good suggestions..

ivansenic commented on sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/service/json/JsonConverter.java:
```diff
@@ -0,0 +1,383 @@
+   * @param numericBooleans `true` if booleans should be treated as numbers
+   * @return the JSON representation of the data
+   */
+  public JsonNode convertToJsonDoc(
```
Let's also add `@WithSpan` here so that we enable tracing for this method..
